yarpl/test/FlowableFlatMapTest.cpp

```diff
@@ -0,0 +1,277 @@
+}
+
+struct CBSubscription : yarpl::flowable::Subscription {
+  template <typename OnReq, typename OnCancel>
+  CBSubscription(OnReq&& r, OnCancel&& c)
+      : onRequest(std::move(r)), onCancel(std::move(c)) {}
+
+  void request(int64_t n) override {
+    onRequest(n);
+  };
+  void cancel() override {
+    onCancel();
+  }
+
+  folly::Function<void(int64_t)> onRequest;
+  folly::Function<void(void)> onCancel;
+};
+
+struct FlowableEvbPair {
+  FlowableEvbPair() = default;
+  Reference<Flowable<int>> flow{nullptr};
+  folly::EventBaseThread evb{};
+};
+
+std::shared_ptr<FlowableEvbPair> make_range_flowable(int start, int end) {
+  auto ret = std::make_shared<FlowableEvbPair>();
+  ret->evb.start("MRF_Worker");
+
+  ret->flow = Flowables::fromPublisher<int>(
+      [&ret, start, end](Reference<Subscriber<int>> s) mutable {
+        auto evb = ret->evb.getEventBase();
+        auto subscription = yarpl::make_ref<CBSubscription>(
+            [=](int64_t req) mutable {
+              /* request */
+              CHECK_EQ(req, 1);
+              if (start >= end) {
+                evb->runInEventBaseThread([=] {
+                  // LOG(INFO) << "sending complete";
+                  s->onComplete();
+                });
+              } else {
+                auto n = start++;
+                evb->runInEventBaseThread([=] {
+                  // LOG(INFO) << "sending n: " << n;
+                  s->onNext(n);
+                });
+              }
+            },
+            /* onCancel: do nothing */
+            []() {});
+
+        evb->runInEventBaseThread([=] { s->onSubscribe(subscription); });
+      });
+
+  return ret;
+}
+
+TEST(FlowableFlatMapTest, Multithreaded) {
+  auto p1 = make_range_flowable(10, 12);
+  auto p2 = make_range_flowable(20, 25);
+
+  folly::Function<Reference<Flowable<int>>(int64_t)> mappingFunc = [&](auto i) {
+    if (i == 0) {
+      return p1->flow;
+    } else {
+      return p2->flow;
+    }
+  };
+
+  auto f = Flowables::range(0, 2)->flatMap(std::move(mappingFunc));
+
+  auto sub = yarpl::make_ref<TestSubscriber<int>>(0);
+  f->subscribe(sub);
+
+  sub->request(2);
+  sub->awaitValueCount(2);
+  EXPECT_TRUE(validate_flatmapped_values(sub->values(), {{10, 11}, {20, 21}}));
+
+  sub->cancel();
+  p1->evb.stop();
+  p2->evb.stop();
+}
+
+TEST(FlowableFlatMapTest, MultithreadedLargeAmount) {
+  auto p1 = make_range_flowable(10000, 40000);
+  auto p2 = make_range_flowable(50000, 80000);
+
+  folly::Function<Reference<Flowable<int>>(int64_t)> mappingFunc = [&](auto i) {
+    if (i == 0) {
+      return p1->flow;
+    } else {
+      return p2->flow;
+    }
+  };
+
+  auto f = Flowables::range(0, 2)->flatMap(std::move(mappingFunc));
+
+  auto sub = yarpl::make_ref<TestSubscriber<int>>();
+  sub->dropValues(true);
+
+  f->subscribe(sub);
+
+  sub->awaitTerminalEvent(std::chrono::seconds{5});
+  EXPECT_EQ(60000, sub->getValueCount());
+  EXPECT_TRUE(sub->isComplete());
+
+  p1->evb.stop();
```

Comment:
Let's don't stop them and see what happens?

Reply:
It needs to stop to flush all the enqueued work, else we're leaking threads. This is effectively 'join'.

Reply:
Another reason they're not scoped: We need a way to manually signal to the EventBaseThread that we want it to flush its work queue and join; else it'll never be destroyed, as the Flowable has a reference to it, and it has a reference to the flowable. Cycle needs to be broken somewhere.